src/peer_select.cc

```diff
@@ -122,79 +145,187 @@ PeerSelector::~PeerSelector()
-    selector->al = al;
+    const char *host = _peer.valid() ? _peer->host : request->url.host();
+    debugs(44, 2, "Find IP destination for: " << request->url << "' via " << host);
+    Dns::nbgethostbyname(host, this);
+}
 
-    selector->entry = entry;
+bool
+PeerSelectionInitiator::wantsMoreDestinations() const
+{
+    const auto maxCount = Config.forward_max_tries;
+    return maxCount >= 0 && foundPaths <
+           static_cast<std::make_unsigned<decltype(maxCount)>::type>(maxCount);
+}
 
-#if USE_CACHE_DIGESTS
 
-    request->hier.peer_select_start = current_time;
+void
+PeerSelectionInitiator::noteIp(const Ip::Address &ip)
+{
+    if (_peer.raw() && !_peer.valid()) //Peer gone, abort?
+        return;
 
-#endif
+    // for TPROXY spoofing, we must skip unusable addresses
+    if (request->flags.spoofClientIp && !(_peer.valid() && _peer->options.no_tproxy) ) {
+        if (ip.isIPv4() != request->client_addr.isIPv4())
+            return; // cannot spoof the client address on this link
+    }
+
+    Comm::ConnectionPointer path = new Comm::Connection();
+    path->remote = ip;
+    path->peerType = _peerType;
+    if (_peerType != PINNED && _peerType != ORIGINAL_DST) {
+        if (_peer.valid()) {
+            path->remote.port(_peer.valid() ? _peer->http_port : request->url.port());
+            path->setPeer(_peer.get());
+        } else
+            path->remote.port(request->url.port());
+    }
 
-    if (selector->entry)
-        selector->entry->lock("peerSelect");
+    // if not pinned check for a configured outgoing address
+    if (_peerType != PINNED)
+        getOutgoingAddress(request.getRaw(), path);
 
-    selector->selectMore();
+    ++foundPaths;
+    noteDestination(path);
+}
+
+void
+PeerSelectionInitiator::noteIps(const Dns::CachedIps *ips, const Dns::LookupDetails &details)
+{
+    if (!ips) {
+        debugs(17, 3, "Unknown host: " << (_peer.valid() ? _peer->host : request->url.host()));
+        // discard any previous error.
+        if (_peerType == HIER_DIRECT) {
+            assert(!_peer.raw());
+            delete lastError;
+            lastError = new ErrorState(ERR_DNS_FAIL, Http::scServiceUnavailable, request.getRaw());
+            lastError->dnsError = details.error;
+            requestMoreDestinations();//nothing to send continue to the next Peer
+        }
+    }
+}
+
+void
+PeerSelectionInitiator::noteLookup(const Dns::LookupDetails &details)
+{
+    request->recordLookup(details);
 }
 
 void
-PeerSelectionInitiator::startSelectingDestinations(HttpRequest *request, const AccessLogEntry::Pointer &ale, StoreEntry *entry)
+PeerSelectionInitiator::startSelectingDestinations(const AccessLogEntry::Pointer &ale, StoreEntry *entry)
 {
     subscribed = true;
-    peerSelect(this, request, ale, entry);
+
+    if (entry)
+        debugs(44, 3, *entry << ' ' << entry->url());
+    else
+        debugs(44, 3, request->method);
+
+    assert(!selector);
+    selector = new PeerSelector(request.getRaw(), entry);
```

Comment:
HttpRequestPointer is an _invasive_ smart pointer. Such smart pointers can be created independently without causing problems (because they all share the reference counting data stored _inside_ the object itself).